### DISCUSSION TITLE
Extension Manifest Proposal

### DISCUSSION THREAD

**Discussion by nanashili** [AuroraEditor/AuroraEditor#769] [Editor]
The below JSON is a manifest proposal for Aurora's extension system. (This is not final and will change over time.)

```json
{
  "name": "Aurora Editor Test Extension",
  "description": "This is a sample extension file that demonstrates how to create a custom editor in Aurora Editor.",
  "icon": "images/mock-debug-icon.png",
  "categories": ["Debuggers"],
  "version": "0.0.1",
  "author": {
    "name": "Wesley de Groot",
    "email": "wdg@auroraeditor.com",
    "company": "Aurora Editor"
  },
  "sponsor": {
    "url": "https://github.com/auroraeditor/auroraeditor"
  },
  "website": "https://auroraeditor.com/?some/long/url/which/does/not/exist",
  "license": "MIT",
  "editor": "0.0.1",
  "permissions": ["FILE_RW", "NOTIFICATIONS", "CUSTOM_EDITOR"],
  "configurations": {
    "debugger": {
      "breakpoints": [{ "language": "markdown" }],
      "supportedLanguages": ["markdown"],
      "id": "markdown",
      "label": "Markdown Debugger",
      "path": "debugger.js",
      "runtime": "node"
    }
  },
  "extensionRequired": ["markdown.editor", "markdown.lsp"],
  "bugs": {
    "url": "https://github.com/auroraeditor/auroraeditor/issues",
    "email": "support@auroraeditor.com"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/auroraeditor/auroraeditor.git"
  },
  "homepage": "https://github.com/auroraeditor/auroraeditort/blob/main/README.md"
}
```

## General Information

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`name`**        | The name of the extension. This will be displayed in the Aurora Editor extension marketplace and other UI elements. |
| **`description`** | A brief description of what the extension does. This helps users understand the purpose and functionality of the extension. |
| **`icon`**        | The path to an icon image that represents the extension. This icon is displayed in the extension marketplace and other UI elements. |
| **`categories`**  | An array of categories that classify the extension. This helps users find the extension based on its functionality. For example, "Debuggers" indicates that this extension provides debugging capabilities. |
| **`version`**     | The version number of the extension. It follows semantic versioning conventions.                  |
| **`website`**     | The URL to the extension's website or documentation page.                                         |
| **`license`**     | The license under which the extension is distributed. For example, "MIT" indicates that the extension is distributed under the MIT License. |
| **`editor`**      | The minimum version of Aurora Editor required to use the extension.                               |
| **`permissions`** | An array of permissions required by the extension. See the [Permissions](#permissions) table for details. |
| **`configurations`** | An object containing configuration settings for the extension. See the [Configurations](#configurations) table for details. |
| **`extensionRequired`** | An array of other extensions required for this extension to function properly.             |
| **`bugs`**        | An object containing information on where to report bugs. See the [Bugs](#bugs) table for details. |
| **`repository`**  | An object containing information about the source code repository for the extension. See the [Repository](#repository) table for details. |
| **`homepage`**    | The URL to the homepage of the extension, typically the README file in the repository.            |

## Author Information

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`author`**      | An object containing information about the author of the extension. See the [Author](#author) table for details. |

### Author

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`name`**        | The name of the author.                                                                          |
| **`email`**       | The email address of the author.                                                                 |
| **`company`**     | The company or organization the author is affiliated with.                                       |

## Sponsor Information

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`sponsor`**     | An object containing information about the sponsor of the extension. See the [Sponsor](#sponsor) table for details. |

### Sponsor

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`url`**         | The URL of the sponsor's webpage.                                                                |

## Permissions

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`permissions`** | An array of permissions required by the extension.                                               |
| `FILE_RW`         | Read and write access to the file system.                                                        |
| `NOTIFICATIONS`   | Permission to send notifications.                                                                |
| `CUSTOM_EDITOR`   | Permission to create a custom editor within Aurora Editor.                                       |

## Configurations

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`configurations`** | An object containing configuration settings for the extension. See the [Debugger](#debugger) table for details. |

### Debugger

| Key                   | Description                                                                                      |
|-----------------------|--------------------------------------------------------------------------------------------------|
| **`breakpoints`**     | An array specifying the languages for which breakpoints are enabled.                             |
| **`supportedLanguages`** | An array of languages for which the debugger is the default.                                   |
| **`id`**              | A unique identifier for the debugger.                                                           |
| **`label`**           | The user-visible name of the debugger.                                                          |
| **`path`**            | The path to the debugger implementation file.                                                   |
| **`runtime`**         | The runtime environment required to run the debugger.                                           |

## Bugs

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`bugs`**        | An object containing information on where to report bugs.                                        |
| `url`             | The URL to the bug reporting page.                                                               |
| `email`           | The email address to report bugs.                                                                |

## Repository

| Key               | Description                                                                                      |
|-------------------|--------------------------------------------------------------------------------------------------|
| **`repository`**  | An object containing information about the source code repository for the extension.             |
| `type`            | The type of version control system (e.g., "git").                                                |
| `url`             | The URL to the repository.                                                                       |

---

**Comment by Angelk90** [comment]
@nanashili : Instead of author I would use authors, there could be more than one contributor to the extension.

**Reply by nanashili** [reply]
Good point, but not everyone might be the original authors there is usually almost always one author that started and then contributions come a long.

**Reply by Angelk90** [reply]
In my opinion, instead of `bugs` it should be `bug`, if it had been an array the nomenclature would have been fine.

**Comment by 0xWDG** [comment]
I have 2 comments about this

```json
{
      "breakpoints": [{ "language": "markdown" }],
      "supportedLanguages": ["markdown"],
}
```

Key | Description
-- | --
breakpoints | An array specifying the languages for which breakpoints are enabled.
supportedLanguages | An array of languages for which the debugger is the default.

----

### 1 Language is redundant and should not be used
Also the table says an `Array` and you are writing an `Object`, I think this is done my mistake.  

```diff
{
-      "breakpoints": [{ "language": "markdown" }],
+      "breakpoints": ["markdown"],
       "supportedLanguages": ["markdown"],
}
```

### 2 breakpoints or breakpointsSupported?
This feels like duplication, `supportedLanguages` should be enough.   
maybe something like `breakpointsSupported: true` (or false)

**Reply by nanashili** [reply]
No it's correct there is nothing wrong with it, breakpoints can be used for multiple file types.

```json
"breakpoints": [
      {
        "language": "javascript"
      },
      {
        "language": "javascriptreact"
      }
    ]
```

Also using 'supportedLangugues' won't work because that's for which files DAP support. Breakpoints would not always be used for all supported files.

**Reply by 0xWDG** [reply]
If breakpoints accept languages, then language is redundant see what you have typed in your own documentation. 
if other things than languages are accepted there is a valid reason to use a dictionary (object) in there otherwise there isn't a valid reason.

**Reply by nanashili** [reply]
Might as well add icon to it for custom icons since the discussion about circular dot or xcodes exist
